chips/nrf52/src/ble_connection/ble_connection_driver.rs

```diff
@@ -0,0 +1,178 @@
+use ble_connection::ble_advertising_hil::RadioChannel;
+use ble_connection::ble_link_layer::LLData;
+use core::fmt;
+
+const NUMBER_CHANNELS: usize = 40;
+const NUMBER_DATA_CHANNELS: usize = NUMBER_CHANNELS - 3;
+
+type ChannelMap = [u8; NUMBER_CHANNELS];
+
+pub struct ConnectionData {
+    last_unmapped_channel: u8,
+    channels: ChannelMap,
+    pub conn_event_counter: u16,
+    hop_increment: u8,
+    number_used_channels: u8,
+    pub aa: u32,
+    pub crcinit: u32,
+    pub transmit_seq_nbr: u8,
+    pub next_seq_nbr: u8,
+    pub conn_interval_start: Option<u32>,
+    pub conn_interval_length_usec: Option<u32>,
+    pub lldata: LLData,
+}
+
+impl PartialEq for ConnectionData {
+    fn eq(&self, other: &ConnectionData) -> bool {
+        self.last_unmapped_channel == other.last_unmapped_channel
+    }
+}
+
+impl Eq for ConnectionData {}
+
+impl fmt::Debug for ConnectionData {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "ConnectionData {{ last_unmapped_channel: {}, conn_event_counter: {}, hop_increment: {}, number_used_channels: {}, aa: {}, crcinit {} }}",
+            self.last_unmapped_channel,
+            self.conn_event_counter,
+            self.hop_increment,
+            self.number_used_channels,
+			self.aa,
+			self.crcinit
+        )
+    }
+}
+
+impl ConnectionData {
+    pub fn new(lldata: LLData) -> ConnectionData {
+        let (channels, number_used_channels) = ConnectionData::expand_channel_map(lldata.chm);
+
+        ConnectionData {
+            last_unmapped_channel: 0,
+            channels,
+            number_used_channels,
+            hop_increment: lldata.hop_and_sca & 0b11111,
+            conn_event_counter: 0,
+            aa: (lldata.aa[0] as u32) << 24 | (lldata.aa[1] as u32) << 16
+                | (lldata.aa[2] as u32) << 8 | (lldata.aa[3] as u32),
+            crcinit: (lldata.crc_init[0] as u32) << 16 | (lldata.crc_init[1] as u32) << 8
+                | (lldata.crc_init[2] as u32),
+            transmit_seq_nbr: 0,
+            next_seq_nbr: 0,
+            conn_interval_start: None,
+            conn_interval_length_usec: None,
+            lldata,
+        }
+    }
+
+    pub fn update_lldata(&mut self, lldata: LLData) {
+        let (channels, number_used_channels) = ConnectionData::expand_channel_map(lldata.chm);
+
+        self.channels = channels;
+        self.number_used_channels = number_used_channels;
+    }
+
+    fn expand_channel_map(chm: [u8; 5]) -> (ChannelMap, u8) {
```

Comment:
`ChannelMap` is esentially a bitmap array (5*8) for each Channel I guess, but why? If this is one of those things you mentioned the other week about saving micro-seconds to get the timing right, I think it's a good idea to document it using comments.

I might've missinterpreted what it's supposed to be, but in that case I think it's even more reason to add some explanation so that it's easier to understand :)

Reply:
This is from the LLData struct sent from the host node in BLE, it contains essentially 40 bitflags that tells the slave which channels should be used in this particular connection. But you're totally right, this could be expressed more clearly in the code with some documentation :) I'll add this to our todo-list!